src/oatpp/core/provider/Pool.hpp

```diff
@@ -223,22 +223,24 @@ class PoolTemplate : public oatpp::base::Countable, public async::CoroutineWaitL
   {
     /* "new" is called directly to keep constructor private */
     auto ptr = std::shared_ptr<PoolTemplate>(new PoolTemplate(provider, maxResources, maxResourceTTL.count()));
-    ptr->startCleanupTask();
+    ptr->startCleanupTask(ptr);
     return ptr;
   }
 
   virtual ~PoolTemplate() {
     stop();
   }
 
-  std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this) {
+  std::shared_ptr<TResource> get(const std::shared_ptr<PoolTemplate>& _this, const std::chrono::duration<v_int64, std::micro>& timeout = std::chrono::microseconds::max()) {
```

Comment:
Its not the problem of being `::max()`. In `libstdc++` this behavior is by design. 

> an object of type std::chrono::duration representing the maximum time to spend waiting. Note that rel_time must be small enough not to overflow when added to std::chrono::steady_clock::now().

And looking at the source of `libstdc++` we can see why:

```c++

template<typename _Rep, typename _Period>
  cv_status
  wait_for(unique_lock<mutex>& __lock, 
                const chrono::duration<_Rep, _Period>& __rtime)
               { return wait_until(__lock, __clock_t::now() + __rtime); }
```

So I assume we just "overflow" when passing `::max()`. 
Therefore we could either catch `::max()` in an if or directly go for `wait_until` and do our own calculation with overflow protection. The last option would also catch other conditions where the passed timeout would overflow.

Reply:
Thanks for the feedback, I'll implement it later today.

I thoght a bit about a way to implement it a different way (without an default value).
Since we most likely have to add extra logic, why don't we provide to public get() functions (one with timeout, one wihtout),
aquire the lock and pass it to an private get function which handles the resource acquisition?
Might be a bit cleaner to read.

Reply:
> However, I think it would also be sufficient to check for `::max()` since I can not imagine an use-case where values greater than a couple of (milli)seconds or `::max()` would be passed.

Yes that's what I thought as well. In that case it would be easy to just have to public get funcitons.

Reply:
> I thoght a bit about a way to implement it a different way (without an default value).
> Since we most likely have to add extra logic, why don't we provide to public get() functions (one with timeout, one wihtout),

*performance wise*  this is the nicer solution. 

> > However, I think it would also be sufficient to check for ::max() since I can not imagine an use-case where values greater than a couple of (milli)seconds or ::max() would be passed.
>
> Yes that's what I thought as well. In that case it would be easy to just have to public get funcitons.

I would wait for @lganzzzo thoughts on that. But I really think this would be sufficient and we could add a warning to the code-doc to not overflow the timeout.

Reply:
I agree overloading might be a better solution than branching with `if`s.

> But I really think this would be sufficient and we could add a warning to the code-doc to not overflow the timeout.

Yes, code-doc is always a good idea.
Also, think we shouldn't worry too much for cases when the user explicitly calls get with `::max()` argument.

The end-user should have the following APIs:

```cpp
/* when there is no resource available */

pool.get();      // blocks forever 
pool.getAsync(); // blocks forever

pool.get(<timeout>);      // blocks until timeout 
pool.getAsync(<timeout>); // blocks until timeout
```

Reply:
Are you speaking about the Pool-functions or the PoolTemplate-functions?
In the other conversation you said:
> Yes, we can make `get(const std::shared_ptr<PoolTemplate>& _this)` and `getAsync(const std::shared_ptr<PoolTemplate>& _this)` static and protected.
> Users won't be affected since they are not supposed to use those functions directly.
> Instead user should use `get()` and `getAsync()`.

The Pool-functions are inherited from core::provider::Provider.
Should we then add new pure virutal functions with timeout parameters in the Provider class, or just add the new overloads to the Pool class. 

In the library scop there would be the following other classes to extend:

- network::virutal_::server::ConnectionProvider
- network::virutal_::client::ConnectionProvider
- network::tcp::server::ConnectionProvider
- network::tcp::client::ConnectionProvider

Reply:
> Are you speaking about the Pool-functions or the PoolTemplate-functions?
In the other conversation you said:

Overloading vs ifs - it about PoolTemplate.

> The Pool-functions are inherited from core::provider::Provider.
Should we then add new pure virutal functions with timeout parameters in the Provider class, or just add the new overloads to the Pool class.

Yes, the new interface should be applied to all providers.

> In the library scop there would be the following other classes to extend:

This is actually a good point. 
While its understood that we have to update providers interface, I dont like the idea 
of having to separate virtual methods with and without timeout.
So the end user interface should have get and getAsync with timeout and a default value

Reply:
Default parameter values for virtual functions can be very complicated:
https://eel.is/c++draft/dcl.fct.default#10

If that's fine, I'll change the Provider interace and add a new defaulted timeout parameter to get and getAsync.

Reply:
Sure, default params should be added to derived classes only.